plugins/aver-hid/fu-aver-safeisp-device.c

```diff
@@ -0,0 +1,448 @@
+			    req,
+			    FU_AVER_SAFEISP_CUSTOM_CMD_UPLOAD_TO_CX3);
+		else if (ISP_SOC == ISP_M12)
+			fu_struct_aver_safeisp_req_set_custom_cmd(
+			    req,
+			    FU_AVER_SAFEISP_CUSTOM_CMD_UPLOAD_TO_M12MO);
+		else
+			return FALSE;
+
+		fu_struct_aver_safeisp_req_set_custom_parm0(req, addr);
+		fu_struct_aver_safeisp_req_set_custom_parm1(req, 512);
+
+		if (!fu_memcpy_safe(req->data,
+				    req->len,
+				    FU_STRUCT_AVER_SAFEISP_REQ_OFFSET_DATA, /* dst */
+				    fu_chunk_get_data(chk),
+				    fu_chunk_get_data_sz(chk),
+				    0x0, /* src */
+				    fu_chunk_get_data_sz(chk),
+				    error))
+			return FALSE;
+
+		/* resize the last packet */
+		if ((i == (fu_chunk_array_length(chunks) - 1)) &&
+		    (fu_chunk_get_data_sz(chk) < 512)) {
+			fu_byte_array_set_size(req, 12 + fu_chunk_get_data_sz(chk), 0x0);
+			fu_struct_aver_safeisp_req_set_custom_parm1(req, fu_chunk_get_data_sz(chk));
+		}
+		if (!fu_aver_safeisp_device_transfer(self, req, res, error))
+			return FALSE;
+		if (!fu_struct_aver_safeisp_res_validate(res->data, res->len, 0x0, error))
+			return FALSE;
+
+		addr += 512;
+
+		/* update progress */
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_upload_checksum(FuAverSafeispDevice *self,
+				       gsize param0,
+				       gsize param1,
+				       GError **error)
+{
+	g_autoptr(GByteArray) req = fu_struct_aver_safeisp_req_new();
+	g_autoptr(GByteArray) res = fu_struct_aver_safeisp_res_new();
+
+	fu_struct_aver_safeisp_req_set_custom_cmd(
+	    req,
+	    FU_AVER_SAFEISP_CUSTOM_CMD_UPLOAD_COMPARE_CHECKSUM);
+	fu_struct_aver_safeisp_req_set_custom_parm0(req, param0);
+	fu_struct_aver_safeisp_req_set_custom_parm1(req, param1);
+	if (!fu_aver_safeisp_device_transfer(self, req, res, error))
+		return FALSE;
+	if (!fu_struct_aver_safeisp_res_validate(res->data, res->len, 0x0, error))
+		return FALSE;
+	if (fu_struct_aver_safeisp_req_get_custom_cmd(res) != FU_AVER_SAFEISP_ACK_STATUS_SUCCESS)
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_update(FuAverSafeispDevice *self, gsize param0, gsize param1, GError **error)
+{
+	g_autoptr(GByteArray) req = fu_struct_aver_safeisp_req_new();
+
+	fu_struct_aver_safeisp_req_set_custom_cmd(req, FU_AVER_SAFEISP_CUSTOM_CMD_UPDATE_START);
+	fu_struct_aver_safeisp_req_set_custom_parm0(req, param0);
+	fu_struct_aver_safeisp_req_set_custom_parm1(req, param1);
+	if (!fu_aver_safeisp_device_transfer(self, req, NULL, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAverSafeispDevice *self = FU_AVER_SAFEISP_DEVICE(device);
+	g_autoptr(FuArchive) archive = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) cx3_fw = NULL;
+	g_autoptr(GBytes) m12_fw = NULL;
```

Comment:
normally all the g_autoptr() ones are listed last in the variable block, and sorted by name.